### ISSUE TITLE
extend DataGeometry objects to provide access to "transform" and "inverse" methods

### ISSUE THREAD

**Issue by jeremymanning** [ContextLab/hypertools#161]
The new `DataGeometry` objects allow users to apply the same sequences of operations (that were applied to old data) to new data.  However, the parameters of those transformations are currently fit anew for the new data.  Therefore it is still inelegant to map new data onto the same space as old data without refitting the model.

The fundamental challenge is that it is not always a straightforward process to fit the transformation parameters and then map new data onto a lower dimensional space.  For example, the scikit learn dimensionality reduction functions (which `hypertools.reduce` wraps) don't have fully consistent APIs.

The two basic functions that would be useful to expose are:
- `model.transform`: use a fitted model to map new data onto a lower dimensional space
- `model.inverse`: take a low-dimensional representation and map it back onto the original high-dimensional feature space

In the long run, it would be nice to support both of these operations for all of the dimensionality reduction models we support in `hypertools.reduce`.  However, as a compromise that might be relatively easy to implement, I suggest that we extend the `DataGeometry` object to include `transform` and `inverse` functions.  We should (initially) only support the simplest cases where:
- No normalization is applied
- No alignment is applied
- `transform` is set to either `model.transform` (if the `transform` function is supported for that model in scikit-learn) or a null "identity" function that just returns whatever data it's passed and outputs a warning message (if the `transform` function is unsupported for that model in scikit-learn, or if normalization or alignment have been applied to the data)
- `inverse` is set to either `model.inverse_transform` (if supported for that model in scikit-learn) or the null identity function (if the `inverse_transform` function is unsupported for that model in scikit-learn, or if normalization or alignment have been applied to the data)

I've compiled a list of which functions support the `transform` and `inverse_transform` methods: [[LINK](https://docs.google.com/spreadsheets/d/1uH43wpOJM0y3tezQm9TsHGKh-8ip_74q3SBiomGsLl0/edit?usp=sharing)]

Eventually we can also support normalized data (by saving the normalization parameters) and aligned data (by saving the alignment parameters).  Essentially we need to save enough so that we can invert those transformations.  I don't think this would be fundamentally difficult, but we just need to think through the right way to implement it.

**Comment by jeremymanning** [comment]
[This paper](http://proceedings.mlr.press/v5/maaten09a/maaten09a.pdf?q=parametric-embedding-for-class-visualization) introduces "parametric t-SNE" that can be used to transform new data (using the same transformation that was learned from training data).  There's no scikit-learn implementation, but the paper includes a link to a MATLAB implementation (that we could potentially translate into Python): http://ticc.uvt.nl/∼lvdrmaaten/tsne

**Comment by andrewheusser** [comment]
looks like he has a python implementation on this page: https://lvdmaaten.github.io/tsne/

**Comment by jeremymanning** [comment]
I can't see how to transform new data (in either the python or matlab versions), but I haven't gone through the paper in detail.

I found the paper above through this discussion: https://www.researchgate.net/post/How_to_integrate_new_data_in_a_TSNE_map

One suggestion from that discussion is to "train a multivariate regressor to predict the map location from the input data."  That's at least straightforward to implement...we could see how it works in practice.

BTW, solving the `transform` issue is going to be important for supporting streaming data-- I think we'll end up wanting to train the mapping onto a low-D space using some initial data, and then we can apply that mapping to new data.

And solving the `inverse` issue will be important for doing analyses like the Raiders analysis from our paper entirely within the toolbox-- i.e. applying a (different) set of transformations to two datasets, and then using the inverse of one of those sets of transformations to map one dataset back onto the original high-dimensional space of the other dataset.

**Comment by jeremymanning** [comment]
Status in the 1.0 update (PR #272 https://github.com/ContextLab/hypertools/pull/272): `DataGeometry` chaining is gone (`plot()` returns a Figure, `load()` returns raw data), but the exact capability requested — fit once, then `model.transform`/`model.inverse_transform` new data without refitting — already works via `hypertools.apply_model(X, 'PCA', ndims=3, return_model=True)`, which returns the real fitted scikit-learn estimator for any model in the registry (`core.model.supported_models()`). Verified live: `model.transform(X_new)` and `model.inverse_transform(result)` both work and return correctly-shaped arrays.

**How this could be built on the 1.0 architecture:** The remaining gap is that this isn't exposed from the higher-level `hyp.reduce()`/`hyp.analyze()` functions, only from `apply_model`. Concretely: add a `return_model=False` passthrough kwarg to `hypertools/reduce/reduce.py`'s `reduce()` (and thread it through `hypertools/tools/analyze.py`'s `analyze()`), returning `(x_reduced, model)` when set — `apply_model` already implements the fit/return-model machinery in `hypertools/core/model.py`, so `reduce()` could simply delegate to it, or just also return the fitted model object it already constructs internally. No new architecture needed; this is a thin API-surface change.

**Alternatives / prior art:**
- `openTSNE` (actively maintained, PyPI v1.0.4 as of this check) implements out-of-sample `.transform()` for t-SNE embeddings specifically — the one model family in hypertools' registry (`TSNE`) that sklearn's own implementation can't do this for. Could be offered as an optional TSNE backend for users who need transform support.
- The issue's own suggestion (translating a MATLAB "parametric t-SNE" implementation) is superseded by openTSNE and by UMAP (already a hypertools dependency), both of which support transforming new data onto an existing embedding.

Note: the stretch goal of also inverting normalization/alignment parameters (saving those transformation params to invert them later) is not implemented and would need its own design — no saved-parameter mechanism exists today. That part is medium/high effort and arguably deserves a separate issue.

Tagging as **low effort** (for the core ask: exposing `return_model` from `reduce()`/`analyze()`); leaving open for consideration in a future release.

**Comment by jeremymanning** [comment]
Also related to issue #227

**Comment by jeremymanning** [comment]
Addressed in PR #272 (HyperTools 1.0). Related to and implemented alongside #227.

`DataGeometry`-style chaining is gone (`plot()` returns a `Figure`, `load()` returns raw data), but the requested capability — fit once, then apply the same fitted transform/inverse-transform to new data without refitting — now works via the public `hyp.Pipeline` class plus `return_model=True` on every dispatcher (`reduce`, `align`, `cluster`, `normalize`, `analyze`, `manip`). A single-stage call returns the fitted model itself; a multi-stage call returns a fitted `Pipeline`. `Pipeline.transform()`/`Pipeline.inverse_transform()` reuse the fitted state — verified via poison-pill tests that monkeypatch `.fit`/`.fit_transform` to raise if called during reuse.

Implemented in `hypertools/core/pipeline.py` (new `Pipeline`/`build_pipeline`, commit `b8ac8fe0`) and wired into `plot()`/`analyze()` via a new `pipeline=` kwarg (commit `b49e70c2`).

```python
import hypertools as hyp

result, model = hyp.apply_model(X, 'PCA', ndims=3, return_model=True)
new_result = model.transform(Y)          # reuses the fit, no refit
recovered = model.inverse_transform(new_result)

# multi-stage, then reapply the whole pipeline to new data via hyp.plot:
_, pipeline = hyp.analyze(A, manip='Smooth', reduce='PCA', align='HyperAlign', return_model=True)
fig = hyp.plot(B, pipeline=pipeline)
```

`Pipeline.inverse_transform` round-trips losslessly through PCA (`n_components=4`, `atol=1e-8`) and raises `NotImplementedError` naming the specific non-invertible step blocking it. 23 tests in `tests/test_pipeline.py`, plus the acceptance scenario in `tests/test_cross_module_kwargs.py` (`analyze(..., return_model=True)` → `plot(B, pipeline=p)`, poison-pilled).

Full suite: ~1275 tests passing.
